### PR TITLE
Multi-Version Mastery and Sniper Scoring

### LINKED ISSUE #109: Different Mastery progressions for different games
All games currently have the same mastery items as H3 on Peacock. However, H1, H2 and H3 do not have the same mastery progression on every location. E.g. Hawkes Bay Lvl. 5 should not have Tanto in H2. 

Originally tracked in #90 .

### PR DESCRIPTION
**Pull request by AnthonyFuller**
This PR does a lot of things, and was originally only meant to be for Multi-Version mastery, but I kept doing things on this branch and didn't want to separate it. Closes #109.

This PR:
- Adds multi-version mastery and subpackages
- Improves H2016 support
    - Normal and pro1 mastery is now supported
    - The destinations screen is now functional (same for H2)
    - pro1 and loadout unlocks have been added to allunlockables (they're only on the server)
- Adds sniper specific unlockables (challenge multipliers)
- Fixes some scpc issues
- Fixes some other various issues
- Adds Sniper Scoring
- Versioned user profiles have been added, alongside documentation for the format
- pro1 is now locked when it should be

I may have missed something, if you need clarification on what a change is for, feel free to ask.

What needs to be done before merging:
- [x] Fix pro1 mastery being available no matter what
- [x] Versioned user profiles to update it to the new format
- [x] Sniper Scoring (requires #273 **to be merged into this branch**)
- **Deferred to be tested for later** Sniper co-op needs to be tested for sniper scoring.
- **Unable to re-create, won't fix in this PR** Fix Freelancer crashing when someone has L100 mastery with no previous CPD data.
- **Deferred to v7** Give people max mastery when they disable mastery progression (see below)

The reason I believe we should give them max mastery is (although it cannot be easily reversed right now) it would simplify the logic for mastery a lot.

### CODE REVIEW

**Comment by AnthonyFuller** [comment]
Versioned user profiles has been added, I'd like extra scrutiny to be applied when reviewing those changes.

**Review by moonysolari — CHANGES_REQUESTED** [review]
ICA Facility now has a "Mastery" page on H2/H3 but it shouldn't. On H1 it will cause a disconnection.

**Comment by AnthonyFuller** [comment]
I thought I fixed that.

**Comment by AnthonyFuller** [comment]
That should be a given, we've never guaranteed backwards compatibility with versions and user profiles.

What's left to be done is validation of co-op which I am almost done with.

**Comment by LennardF1989** [comment]
But with the profile migration system in place, we should support it IMO. We can't constantly let people wipe their profiles no that progression is a thing.

**Comment by AnthonyFuller** [comment]
Re-requested reviews. Good to merge now, co-op not tested, doing that at some point before release.

File: components/menus/destinations.ts [hunk 1]
```diff
@@ -149,8 +159,44 @@ export function destinationsMenu(req: RequestWithJwt): GameFacingDestination[] {
...
+                                      "mission",
+                                      "pro1",
+                                  ),
+                              },
+                          }
+                        : undefined,
             },
         }
+
+        // TODO: THIS IS NOT CORRECT FOR 2016!
```
Comment (LennardF1989) [inline components/menus/destinations.ts:190]: Should we make a new issue in order to not forget?

File: components/candle/progressionService.ts [hunk 2]
```diff
@@ -39,28 +42,38 @@ export class ProgressionService {
         dropIds: string[],
         contractSession: ContractSession,
         userProfile: UserProfile,
+        gameVersion: GameVersion,
```
Comment (LennardF1989) [inline components/candle/progressionService.ts:45]: gameVersion is already in the ContractSession.

Comment (LennardF1989) [reply components/candle/progressionService.ts:45]: Same for everything down this path.

File: components/evergreen.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ import { ContractProgressionData } from "./types/types"
 import { getFlag } from "./flags"
 import { EVERGREEN_LEVEL_INFO } from "./utils"
 
-export async function setCpd(
+export function setCpd(
```
Comment (LennardF1989) [inline components/evergreen.ts:25]: Why was this done? EDIT: Nvm. writeUserData is not async either.

File: components/index.ts [hunk 4]
```diff
@@ -91,6 +88,8 @@ import { liveSplitManager } from "./livesplit/liveSplitManager"
 import { cheapLoadUserData } from "./databaseHandler"
 import { reportRouter } from "./contracts/reportRouting"
 
+loadFlags()
```
Comment (LennardF1989) [inline components/index.ts:91]: Why has this been moved, it was purposely one of the first things so you can use flags immediately. We have to be VERY careful that nothing in the imports try to use flags in their global logic.

**Review by LennardF1989 — CHANGES_REQUESTED** [review]
Please see my questions.

File: components/index.ts (hunk 4 above)
Comment (AnthonyFuller) [reply components/index.ts:91]: I didn't consciously move it, so it was probably Webstorm. I've moved it back, shouldn't have been an issue though.

File: components/menus/destinations.ts (hunk 1 above)
Comment (AnthonyFuller) [reply components/menus/destinations.ts:190]: Sure.

File: components/candle/progressionService.ts (hunk 2 above)
Comment (AnthonyFuller) [reply components/candle/progressionService.ts]: Fixed.

File: components/index.ts (hunk 4 above)
Comment (RDIL) [reply components/index.ts]: Yeah, i’ve been meaning to file a YouTrack issue soon

**Review by LennardF1989 — APPROVED** [review]
LGTM

File: components/candle/challengeService.ts [hunk 5]
```diff
@@ -354,7 +354,7 @@ export abstract class ChallengeRegistry {
      * Parse a challenge's context listeners into the format used internally.
      *
      * @param challenge The challenge.
-     * @param Context The context to use for parsing.
+     * @param Context? the current context of the challenge.
```
Comment (RDIL) [inline components/candle/challengeService.ts:357]:
```suggestion
     * @param Context The current context of the challenge.
```

File: components/index.ts [hunk 6]
```diff
@@ -376,6 +319,62 @@ app.use(
         }),
 )
 
+app.get(
```
Comment (RDIL) [inline components/index.ts:322]: I really hope this doesn't break scpc

**Review by RDIL — COMMENTED** [review]
This is gonna kill perf!

File: components/index.ts (hunk 6 above)
Comment (AnthonyFuller) [reply components/index.ts:322]: Correction: it **FIXES** it. Someone broke it.

**Review by RDIL — APPROVED** [review]

**Pull request merged**